### ISSUE TITLE
Clang: ABI incompatibility with MSVC on i686-windows for struct containing overaligned type

### ISSUE THREAD

**Issue by erikdesjardins** [llvm/llvm-project#63874]
Somewhat related to #63257.

In https://reviews.llvm.org/D72114, a special case was added to pass overaligned arguments indirectly, to match MSVC. However, it doesn't quite match MSVC behavior. MSVC seems to apply the special case transitively, if any field of the struct contains an overaligned type, whereas Clang only applies the special case when the toplevel struct itself is overaligned.

For example, for the following struct definitions:

```c
__declspec(align(8))
struct Align
{
    uint64_t a;
    uint64_t b;
};

struct Wrap1Align
{
    struct Align a;
    uint32_t b;
};

struct Wrap2Align {
    struct Wrap1Align a;
    uint8_t b;
};
```

MSVC passes all three indirectly, whereas Clang only passes `Align` indirectly.

Demo: https://godbolt.org/z/766G9zx53

**Comment by nikic** [comment]
If I'm reading this right, it works correctly on trunk: https://godbolt.org/z/bs93rWeEb Possibly fixed as a side-effect of https://github.com/llvm/llvm-project/commit/651e5ae62d29fdb07eb85f75ab7e686b98479f3a?

**Comment by erikdesjardins** [comment]
Yep, it looks fixed, I should've checked trunk. No change required then, but we should add a test for this case, since it seems not to be covered.
